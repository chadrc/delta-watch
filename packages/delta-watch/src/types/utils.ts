import {ObjectWatcher} from "../core/ObjectWatcher";

export interface DeltaWatchInternals {
  watcher: ObjectWatcher
  type: string
}

export function makeAccessorHandler(mutatorMethods: string[]) {
  return {
    get: function (obj: Date, prop: PropertyKey) {
      if (prop in obj) {
        let field = (obj as any)[prop];
        if (typeof field === 'function') {
          if (mutatorMethods.indexOf(prop as string) !== -1) {
            throw Error("Cannot access a mutator method on an Accessor object.");
          }
          field = field.bind(obj);
        }
        return field;
      }
    },
    set: function () {
      throw Error("Cannot set a value on the Accessor object.");
    }
  };
}

export type CanSetPropFunc = (prop: PropertyKey) => boolean;

export function makeMutationHandler<T extends object>(internals: DeltaWatchInternals,
                                                      mutatorMethods: string[],
                                                      canSetProp: boolean | CanSetPropFunc = false ): ProxyHandler<T> {
  return {
    get: function (_: T, prop: PropertyKey) {
      if (prop === "__DeltaWatchInternals") {
        return internals;
      }

      let fieldMutator = ObjectWatcher.getMutator(internals.watcher, prop);
      let fieldWatcher = ObjectWatcher.getFieldWatcher(internals.watcher, prop);
      if (prop in internals.watcher._data) {
        let field = (internals.watcher._data as any)[prop];
        if (typeof field === 'function') {
          field = field.bind(internals.watcher._data);
          if (mutatorMethods.indexOf(prop as string) !== -1) {
            return function (...args: any[]) {
              let result = field(...args);
              internals.watcher._notifySubscribers(true, true, true);
              return result;
            }
          }
        } else {
          // let fieldMutator = ObjectWatcher.getMutator(self._watcher, field);
          // return fieldMutator == null ? self._watcher._properties[field]._data : fieldMutator;

          if (fieldMutator !== null && typeof fieldMutator !== 'undefined') {
            return fieldMutator;
          } else if (fieldWatcher !== null && typeof fieldWatcher !== 'undefined') {
            internals.watcher._makeMutator(prop);
            return ObjectWatcher.getMutator(internals.watcher, prop);
          }
        }
        return field;
      } else if (fieldMutator !== null && typeof fieldMutator !== 'undefined') {
        return fieldMutator;
      } else if (fieldWatcher !== null && typeof fieldWatcher !== 'undefined') {
        internals.watcher._makeMutator(prop);
        return ObjectWatcher.getMutator(internals.watcher, prop);
      }
    },
    set: function (_: T, prop: PropertyKey, value: any): boolean {
      if (prop === "__DeltaWatchInternals") {
        throw Error("Cannot set value of __DeltaWatchInternals");
      }

      let willSetProp;
      if (canSetProp instanceof Function) {
        willSetProp = canSetProp(prop);
      } else {
        willSetProp = canSetProp;
      }
      // If any prop should be accepted as settable
      // or settableProps contains the given prop
      if (willSetProp === true) {
        // Need to mutate before array check,
        // because watcher's make mutator uses current value to determine
        // to make Object or Array mutator

        // if data is undefined then we need to make it an object
        if (typeof internals.watcher._data === 'undefined') {
          internals.watcher._parent._data[internals.watcher._parentKey] = {};
        }
        internals.watcher._data[prop] = value;
        let fieldMutator = ObjectWatcher.getMutator(internals.watcher, prop);
        let type = internals.watcher._typeRegistry.getTypeForValue(value);

        if (Array.isArray(value) &&
          (fieldMutator === null || typeof fieldMutator === 'undefined'
            || (fieldMutator as any).__DeltaWatchInternals.type !== type)) {
          // if mutator hasn't been initialized
          // or the type of the mutator doesn't match the new value
          // make a new mutator for it
          internals.watcher._makeMutator(prop);
        }

        let watcherProperties = internals.watcher._properties[prop];
        if (watcherProperties) {
          watcherProperties._notifySubscribers(true, true);
        }
        return true;
      }

      throw Error(`${prop} not settable on Mutator of type ${internals.type}`)
    }
  };
}