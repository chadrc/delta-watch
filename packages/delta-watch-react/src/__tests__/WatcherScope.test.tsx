import 'jest';
import * as React from 'react';
import DeltaWatchReact from '../index';
import * as renderer from 'react-test-renderer';
import './setup'
import {flushTimers} from "./utils";

interface User {
  name: string;
  email: string;
}

describe(`WatcherScope`, () => {
  it(`WatchStore receives scope watcher instead of root`, () => {
    const {WatchStore, MakeWatcherScope, Store} = DeltaWatchReact.MakeStore({
      user: {
        name: "John Doe",
        email: "johndoe@example.com"
      }
    });

    // Make a scope to the user object
    let {Scope, withScope} = MakeWatcherScope((watcher: any) => watcher.user);

    // Component that will be rendered
    const ChildComp = (props) => (
      <span>{props.name} - {props.email}</span>
    );

    // Watch store, watcher will be the user watcher not the root store watcher
    const WatchingComp = WatchStore((watcher: User) => ({
      name: watcher.name,
      email: watcher.email
    }))(ChildComp);

    // Wrapper component to render with Scope context
    const ScopedComp = withScope(WatchingComp);

    // Make and test final rendering
    let Parent = () => (
      <Scope>
        <ScopedComp/>
      </Scope>
    );

    let component = renderer.create(<Parent/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    Store.Mutator.user = {
      name: "Jane Smith",
      email: "janesmith@example.com"
    };

    flushTimers();

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});