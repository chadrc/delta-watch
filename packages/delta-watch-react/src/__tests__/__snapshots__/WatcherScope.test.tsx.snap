// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WatcherScope WatchStore receives scope watcher instead of root 1`] = `
<span>
  John Doe
   - 
  johndoe@example.com
</span>
`;

exports[`WatcherScope WatchStore receives scope watcher instead of root 2`] = `
<span>
  Jane Smith
   - 
  janesmith@example.com
</span>
`;
